### ISSUE TITLE
SPV: Out of order offsets on block members

### ISSUE THREAD

**Issue by yavn** [KhronosGroup/glslang#541]
Given that we agreed to accept out of order block member offsets (in Vulkan), I wanted to test this:

```
layout(set = 0, binding = 0, std140) uniform Data {
    layout(offset = 8)  int a;
    layout(offset = 4)  int b;
    layout(offset = 0)  int c;
    layout(offset = 12) int d;
} u_data;
```

But currently it doesn't seem to work and I'm getting "offset cannot lie in previous members" error.

I wanted to add support for this in Vulkan mode only, but I'm not sure how a mix of explicit and implicit offsets should be handled:

```
layout(set = 0, binding = 0, std140) uniform Data {
    layout(offset = 8)  int a;
    layout(offset = 4)  int b;
                        int c;  // where should I go?
    layout(offset = 12) int d;
} u_data;
```

`KHR_vulkan_glsl` says about this:

> "If offset was declared, start with that offset, otherwise start with the offset immediately following the preceding member (in declaration order)."

I think this rule may be a bit too simplistic. The following offset after `b` would be 8, which collides with member `a`. Am I right?

We probably want to put that member at the largest unused offset (in declaration order), so at 12. In this case that would collide with the explicitly offset member `d`. To catch that error at compile-time, glslang would now have to keep track of used offsets (and sizes), which adds a bit of overhead.

I'd be happy to propose a PR for this, but wanted to confirm what is the expected behavior. Thanks.

**Comment by ratchetfreak** [comment]
A simple rule like the one in KHR_vulkan_glsl is much easier to reason about and doesn't surprise the user. Starting with keeping track of unused space in the data layout is overhead that's not needed for all parties.

With the rule as defined in KHR_vulkan_glsl then in the second `Data`  `a` and `c` will collide and should error. You only need to keep track of all the members and don't have to worry about unused space, just check all new fields against all existing fields.

**Comment by johnkslang** [comment]
The rule is intentionally simple.  @ratchetfreak stated appropriate reasoning.  That is, from the intended perspective, in the @yavn's second example, the shader requested the overlap.

I will look at what the problem is with overlapping offsets.

**Comment by johnkslang** [comment]
This is fixed by 19bdf90eba71390f04bb85226337517df65d73e2.

**Comment by yavn** [comment]
Okay, that's reasonable. So the only missing part is to give an error when such overlap happens.

The fix works for me, I can progress with that. Thanks.
